### PR TITLE
ENH: HiGHS submodule and update

### PR DESCRIPTION
**Pull request by mckib2**
#### What does this implement/fix?

- remove HiGHS source tree from the SciPy repository
- create and use a fork of the official HiGHS repository for easier upstream/downstream change management
- modifications to Cython wrappers to catch up to the current HiGHS API
    - includes changes to pxd files to add/update data members and class interfaces

#### Additional information

Inclusion of submodules in SciPy was not yet standard practice when HiGHS was first integrated.  It was discussed but ultimately a vendor approach won out.  With the inclusion of Boost, UNURAN, and PROPACK submodules, it makes sense to move HiGHS to a submodule as well, especially given the development pace of the project and the unreasonable review burden that would be put on SciPy maintainers to keep an up-to-date version.

Blockers:
- Auto write of High.log to working directory, see https://github.com/ERGO-Code/HiGHS/issues/586
- Moving https://github.com/mckib2/highs into SciPy's org

### CODE REVIEW

**Comment by mckib2** [comment]
@mdhaber Building and passing tests on Linux, will try Windows/Mac hopefully tomorrow evening.  Please take a look and suggest how we can make this as easy as possible for SciPy reviewers

**Comment by mdhaber** [comment]
> Please take a look and suggest how we can make this as easy as possible for SciPy reviewers

Deletions make sense, but maybe add brief comments in the PR about what the new code is for?

File: scipy/optimize/_highs/cython/src/HConst.pxd [hunk 1]
```diff
@@ -6,43 +6,47 @@ from libcpp.string cimport string
...
-    const double HIGHS_CONST_TINY
-    const double HIGHS_CONST_ZERO
-    const int HIGHS_THREAD_LIMIT
-    const bool allow_infinite_costs
-    const string FILENAME_DEFAULT
+    const int HIGHS_CONST_I_INF "kHighsIInf"
+    const double HIGHS_CONST_INF "kHighsInf"
+    const double kHighsTiny
+    const double kHighsZero
+    const int kHighsThreadLimit
```
Comment (mdhaber) [inline scipy/optimize/_highs/cython/src/HConst.pxd:13]: For instance, maybe it's obvious to someone with more experience, but why the changes here?

Comment (mckib2) [reply scipy/optimize/_highs/cython/src/HConst.pxd:13]: Yep, I can try to describe this a little better.  These fall under the broad category of aligning Cython wrappers with the current HiGHS API.  The quoted string is the literal that Cython will use in generated code and matches the changes to HiGHS' API.  I'm taking advantage of that here as a poor-man's typedef to make fewer changes to the existing pxd files

**Comment by mckib2** [comment]
@mdhaber I've struggled for a few hours now trying to get Windows to run for me.  It's building fine with the latest changes, but I get the following error when trying to run tests:
```
ImportError while importing test module 'C:\Users\nmckibben\Documents\scipy\scipy\optimize\tests\test__linprog_clean_inputs.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
C:\Python310\lib\importlib\__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
scipy\optimize\__init__.py:399: in <module>
    from ._optimize import *
scipy\optimize\_optimize.py:33: in <module>
    from scipy.sparse.linalg import LinearOperator
scipy\sparse\linalg\__init__.py:120: in <module>
    from ._isolve import *
scipy\sparse\linalg\_isolve\__init__.py:6: in <module>
    from .lgmres import lgmres
scipy\sparse\linalg\_isolve\lgmres.py:7: in <module>
    from scipy.linalg import get_blas_funcs
scipy\linalg\__init__.py:198: in <module>
    from ._misc import *
scipy\linalg\_misc.py:4: in <module>
    from .lapack import get_lapack_funcs
scipy\linalg\lapack.py:818: in <module>
    from scipy.linalg import _flapack
E   ImportError: DLL load failed while importing _flapack: The operating system cannot run %1.
``` 

https://github.com/scipy/scipy/issues/13717 seems related -- I'm not building with `clang-cl` or Pythran and I'm getting a different error than reported there, but maybe the compiler is the issue.  I'm not sure at this point.  I always forget how fun it is to set Windows up for the first time...

**Comment by mdhaber** [comment]
: ( Sorry to hear that. Do you want me to test whether I get the same error?

**Comment by mckib2** [comment]
> : ( Sorry to hear that. Do you want me to test whether I get the same error?

Sure, that would be helpful.  Any hints on how to get `clang-cl` to be used instead of `cl`?  I have both MSVC BuildTools, mingw, and llvm installed (all through chocolately)

**Comment by mdhaber** [comment]
You probably saw my comments starting [here](https://github.com/scipy/scipy/issues/13717#issuecomment-885358911). I really didn't know what I was doing. My strategy to use one was to force-uninstall the other.

**Comment by mdhaber** [comment]
I have `SCIPY_USE_PYTHRAN=0` right now because ASV was having trouble with it being 1. I get an error about `CMakeLists.txt`.

<details>

```
(scipydev) C:\Users\matth\scipy>git clean -dxf
Skipping repository benchmarks/env/43aced745fbfea41add6be22ac8298b4/project
Removing site.cfg

(scipydev) C:\Users\matth\scipy>python setup.py build_ext --inplace
Cythonizing sources
Running scipy\linalg\_generate_pyx.py
Running scipy\special\_generate_pyx.py
Running scipy\stats\_generate_pyx.py
Processing scipy\cluster\_hierarchy.pyx
Processing scipy\cluster\_optimal_leaf_ordering.pyx
Processing scipy\cluster\_vq.pyx
Processing scipy\fftpack\convolve.pyx
Processing scipy\interpolate\interpnd.pyx
Processing scipy\interpolate\_bspl.pyx
Processing scipy\interpolate\_ppoly.pyx
Processing scipy\io\matlab\_mio5_utils.pyx
Processing scipy\io\matlab\_mio_utils.pyx
Processing scipy\io\matlab\_streams.pyx
Processing scipy\linalg\cython_blas.pyx
Processing scipy\linalg\cython_lapack.pyx
Processing scipy\linalg\_cythonized_array_utils.pyx
Processing scipy\linalg\_decomp_update.pyx.in
Processing scipy\linalg\_matfuncs_sqrtm_triu.pyx
Processing scipy\linalg\_solve_toeplitz.pyx
Processing scipy\ndimage\src\_cytest.pyx
Processing scipy\ndimage\src\_ni_label.pyx
Processing scipy\optimize\_bglu_dense.pyx
Processing scipy\optimize\_group_columns.pyx
Processing scipy\optimize\cython_optimize\_zeros.pyx.in
Processing scipy\optimize\tnc\moduleTNC.pyx
Processing scipy\optimize\_highs\cython\src\_highs_constants.pyx
Processing scipy\optimize\_highs\cython\src\_highs_wrapper.pyx
Processing scipy\optimize\_lsq\givens_elimination.pyx
Processing scipy\optimize\_trlib\_trlib.pyx
Processing scipy\signal\_max_len_seq_inner.pyx
Processing scipy\signal\_peak_finding_utils.pyx
Processing scipy\signal\_sosfilt.pyx
Processing scipy\signal\_spectral.pyx
Processing scipy\signal\_upfirdn_apply.pyx
Processing scipy\sparse\_csparsetools.pyx.in
Processing scipy\sparse\csgraph\_flow.pyx
Processing scipy\sparse\csgraph\_matching.pyx
Processing scipy\sparse\csgraph\_min_spanning_tree.pyx
Processing scipy\sparse\csgraph\_reordering.pyx
Processing scipy\sparse\csgraph\_shortest_path.pyx
Processing scipy\sparse\csgraph\_tools.pyx
Processing scipy\sparse\csgraph\_traversal.pyx
Processing scipy\spatial\_ckdtree.pyx
Processing scipy\spatial\_hausdorff.pyx
Processing scipy\spatial\_qhull.pyx
Processing scipy\spatial\_voronoi.pyx
Processing scipy\spatial\transform\_rotation.pyx
Processing scipy\special\cython_special.pyx
Processing scipy\special\_comb.pyx
Processing scipy\special\_ellip_harm_2.pyx
Processing scipy\special\_test_round.pyx
Processing scipy\special\_ufuncs.pyx
Processing scipy\special\_ufuncs_cxx.pyx
Processing scipy\stats\_biasedurn.pyx
Processing scipy\stats\_qmc_cy.pyx
Processing scipy\stats\_sobol.pyx
Processing scipy\stats\_stats.pyx
Processing scipy\stats\_boost\src\beta_ufunc.pyx
Processing scipy\stats\_boost\src\binom_ufunc.pyx
Processing scipy\stats\_boost\src\hypergeom_ufunc.pyx
Processing scipy\stats\_boost\src\nbinom_ufunc.pyx
Processing scipy\stats\_unuran\unuran_wrapper.pyx
Processing scipy\_lib\messagestream.pyx
Processing scipy\_lib\_ccallback_c.pyx
Processing scipy\_lib\_test_deprecation_call.pyx
Processing scipy\_lib\_test_deprecation_def.pyx
Running from SciPy source directory.
lapack_opt_info:
lapack_mkl_info:
No module named 'numpy.distutils._msvccompiler' in numpy.distutils; trying from distutils
customize MSVCCompiler
  libraries mkl_rt not found in ['C:/Users/matth/Anaconda3/envs/scipydev\\Library\\lib']
  NOT AVAILABLE

openblas_lapack_info:
  libraries openblas not found in ['C:\\opt\\64\\lib']
get_default_fcompiler: matching types: '['gnu', 'intelv', 'absoft', 'compaqv', 'intelev', 'gnu95', 'g95', 'intelvem', 'intelem', 'flang']'
customize GnuFCompiler
Could not locate executable g77
Could not locate executable f77
customize IntelVisualFCompiler
Could not locate executable ifort
Could not locate executable ifl
customize AbsoftFCompiler
Could not locate executable f90
customize CompaqVisualFCompiler
Could not locate executable DF
customize IntelItaniumVisualFCompiler
Could not locate executable efl
customize Gnu95FCompiler
Found executable C:\msys64\mingw64\bin\gfortran.exe
Using built-in specs.
COLLECT_GCC=C:\msys64\mingw64\bin\gfortran.exe
COLLECT_LTO_WRAPPER=C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/lto-wrapper.exe
Target: x86_64-w64-mingw32
Configured with: ../gcc-10.2.0/configure --prefix=/mingw64 --with-local-prefix=/mingw64/local --build=x86_64-w64-mingw32 --host=x86_64-w64-mingw32 --target=x86_64-w64-mingw32 --with-native-system-header-dir=/mingw64/x86_64-w64-mingw32/include --libexecdir=/mingw64/lib --enable-bootstrap --with-arch=x86-64 --with-tune=generic --enable-languages=c,lto,c++,fortran,ada,objc,obj-c++ --enable-shared --enable-static --enable-libatomic --enable-threads=posix --enable-graphite --enable-fully-dynamic-string --enable-libstdcxx-filesystem-ts=yes --enable-libstdcxx-time=yes --disable-libstdcxx-pch --disable-libstdcxx-debug --disable-isl-version-check --enable-lto --enable-libgomp --disable-multilib --enable-checking=release --disable-rpath --disable-win32-registry --disable-nls --disable-werror --disable-symvers --disable-plugin --with-libiconv --with-system-zlib --with-gmp=/mingw64 --with-mpfr=/mingw64 --with-mpc=/mingw64 --with-isl=/mingw64 --with-pkgversion='Rev5, Built by MSYS2 project' --with-bugurl=https://github.com/msys2/MINGW-packages/issues --with-gnu-as --with-gnu-ld --with-boot-ldflags='-pipe -Wl,--dynamicbase,--high-entropy-va,--nxcompat,--default-image-base-high -Wl,--disable-dynamicbase -static-libstdc++ -static-libgcc' 'LDFLAGS_FOR_TARGET=-pipe -Wl,--dynamicbase,--high-entropy-va,--nxcompat,--default-image-base-high' --enable-linker-plugin-flags='LDFLAGS=-static-libstdc++\ -static-libgcc\ -pipe\ -Wl,--dynamicbase,--high-entropy-va,--nxcompat,--default-image-base-high\ -Wl,--stack,12582912'
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 10.2.0 (Rev5, Built by MSYS2 project)
  libraries openblas not found in ['C:\\opt\\64\\lib']
Library openblas was not found. Ignoring
  libraries openblas not found in ['C:\\opt\\64\\lib']
Runtime library openblas was not found. Ignoring
  FOUND:
    library_dirs = ['C:\\Users\\matth\\scipy\\build\\openblas_lapack_info']
    libraries = ['openblas_lapack_info']
    language = f77
    define_macros = [('HAVE_CBLAS', None)]

  FOUND:
    library_dirs = ['C:\\Users\\matth\\scipy\\build\\openblas_lapack_info']
    libraries = ['openblas_lapack_info']
    language = f77
    define_macros = [('HAVE_CBLAS', None)]

blas_opt_info:
blas_mkl_info:
  libraries mkl_rt not found in ['C:/Users/matth/Anaconda3/envs/scipydev\\Library\\lib']
  NOT AVAILABLE

blis_info:
  libraries blis not found in ['C:/Users/matth/Anaconda3/envs/scipydev\\Library\\lib']
  NOT AVAILABLE

openblas_info:
  libraries openblas not found in ['C:\\opt\\64\\lib']
get_default_fcompiler: matching types: '['gnu', 'intelv', 'absoft', 'compaqv', 'intelev', 'gnu95', 'g95', 'intelvem', 'intelem', 'flang']'
customize GnuFCompiler
customize IntelVisualFCompiler
customize AbsoftFCompiler
customize CompaqVisualFCompiler
customize IntelItaniumVisualFCompiler
customize Gnu95FCompiler
Using built-in specs.
COLLECT_GCC=C:\msys64\mingw64\bin\gfortran.exe
COLLECT_LTO_WRAPPER=C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/lto-wrapper.exe
Target: x86_64-w64-mingw32
Configured with: ../gcc-10.2.0/configure --prefix=/mingw64 --with-local-prefix=/mingw64/local --build=x86_64-w64-mingw32 --host=x86_64-w64-mingw32 --target=x86_64-w64-mingw32 --with-native-system-header-dir=/mingw64/x86_64-w64-mingw32/include --libexecdir=/mingw64/lib --enable-bootstrap --with-arch=x86-64 --with-tune=generic --enable-languages=c,lto,c++,fortran,ada,objc,obj-c++ --enable-shared --enable-static --enable-libatomic --enable-threads=posix --enable-graphite --enable-fully-dynamic-string --enable-libstdcxx-filesystem-ts=yes --enable-libstdcxx-time=yes --disable-libstdcxx-pch --disable-libstdcxx-debug --disable-isl-version-check --enable-lto --enable-libgomp --disable-multilib --enable-checking=release --disable-rpath --disable-win32-registry --disable-nls --disable-werror --disable-symvers --disable-plugin --with-libiconv --with-system-zlib --with-gmp=/mingw64 --with-mpfr=/mingw64 --with-mpc=/mingw64 --with-isl=/mingw64 --with-pkgversion='Rev5, Built by MSYS2 project' --with-bugurl=https://github.com/msys2/MINGW-packages/issues --with-gnu-as --with-gnu-ld --with-boot-ldflags='-pipe -Wl,--dynamicbase,--high-entropy-va,--nxcompat,--default-image-base-high -Wl,--disable-dynamicbase -static-libstdc++ -static-libgcc' 'LDFLAGS_FOR_TARGET=-pipe -Wl,--dynamicbase,--high-entropy-va,--nxcompat,--default-image-base-high' --enable-linker-plugin-flags='LDFLAGS=-static-libstdc++\ -static-libgcc\ -pipe\ -Wl,--dynamicbase,--high-entropy-va,--nxcompat,--default-image-base-high\ -Wl,--stack,12582912'
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 10.2.0 (Rev5, Built by MSYS2 project)
  libraries openblas not found in ['C:\\opt\\64\\lib']
Library openblas was not found. Ignoring
  libraries openblas not found in ['C:\\opt\\64\\lib']
Runtime library openblas was not found. Ignoring
  FOUND:
    library_dirs = ['C:\\Users\\matth\\scipy\\build\\openblas_info']
    libraries = ['openblas_info']
    language = f77
    define_macros = [('HAVE_CBLAS', None)]

  FOUND:
    library_dirs = ['C:\\Users\\matth\\scipy\\build\\openblas_info']
    libraries = ['openblas_info']
    language = f77
    define_macros = [('HAVE_CBLAS', None)]

Traceback (most recent call last):
  File "C:\Users\matth\scipy\setup.py", line 622, in <module>
    setup_package()
  File "C:\Users\matth\scipy\setup.py", line 618, in setup_package
    setup(**metadata)
  File "C:\Users\matth\Anaconda3\envs\scipydev\lib\site-packages\numpy\distutils\core.py", line 135, in setup
    config = configuration()
  File "C:\Users\matth\scipy\setup.py", line 528, in configuration
    config.add_subpackage('scipy')
  File "C:\Users\matth\Anaconda3\envs\scipydev\lib\site-packages\numpy\distutils\misc_util.py", line 1016, in add_subpackage
    config_list = self.get_subpackage(subpackage_name, subpackage_path,
  File "C:\Users\matth\Anaconda3\envs\scipydev\lib\site-packages\numpy\distutils\misc_util.py", line 982, in get_subpackage
    config = self._get_configuration_from_setup_py(
  File "C:\Users\matth\Anaconda3\envs\scipydev\lib\site-packages\numpy\distutils\misc_util.py", line 924, in _get_configuration_from_setup_py
    config = setup_module.configuration(*args)
  File "scipy\setup.py", line 18, in configuration
    config.add_subpackage('optimize')
  File "C:\Users\matth\Anaconda3\envs\scipydev\lib\site-packages\numpy\distutils\misc_util.py", line 1016, in add_subpackage
    config_list = self.get_subpackage(subpackage_name, subpackage_path,
  File "C:\Users\matth\Anaconda3\envs\scipydev\lib\site-packages\numpy\distutils\misc_util.py", line 982, in get_subpackage
    config = self._get_configuration_from_setup_py(
  File "C:\Users\matth\Anaconda3\envs\scipydev\lib\site-packages\numpy\distutils\misc_util.py", line 924, in _get_configuration_from_setup_py
    config = setup_module.configuration(*args)
  File "scipy\optimize\setup.py", line 130, in configuration
    config.add_subpackage('_highs')
  File "C:\Users\matth\Anaconda3\envs\scipydev\lib\site-packages\numpy\distutils\misc_util.py", line 1016, in add_subpackage
    config_list = self.get_subpackage(subpackage_name, subpackage_path,
  File "C:\Users\matth\Anaconda3\envs\scipydev\lib\site-packages\numpy\distutils\misc_util.py", line 982, in get_subpackage
    config = self._get_configuration_from_setup_py(
  File "C:\Users\matth\Anaconda3\envs\scipydev\lib\site-packages\numpy\distutils\misc_util.py", line 924, in _get_configuration_from_setup_py
    config = setup_module.configuration(*args)
  File "scipy\optimize\_highs\setup.py", line 64, in configuration
    _major_dot_minor = _get_version(
  File "scipy\optimize\_highs\setup.py", line 51, in _get_version
    with open(CMakeLists, 'r', encoding='utf-8') as f:
FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\matth\\scipy\\scipy\\_lib\\highs\\CMakeLists.txt'
```

</details>

**Comment by mckib2** [comment]
@mdhaber Did you pull the new highs submodule?  Take a look in scipy/_lib/highs to make sure it's not empty

**Comment by mdhaber** [comment]
Hah of course not : P
No errors for me!

**Comment by mckib2** [comment]
Great news, though I'm still not getting anywhere on my Windows build.  I'm going to try uninstalling everything and starting fresh with Anaconda

And that's good that you ran into that empty submodule -- there should be a more descriptive error

**Comment by mckib2** [comment]
Just verified everything runs on MacOS and for now I trust you that it works on Windows.  Are we ready to open a PR against the main scipy repo?

**Comment by mdhaber** [comment]
Almost. Just review the diff for unintended changes (e.g. there are changes to stats at the bottom that don't look related) first. Then when you post, please add those comments about changes (other than deletions) and why they are necessary. I'd avoid complicating this PR with unnecessary changes.

File: .github/workflows/linux.yml [hunk 2]
```diff
@@ -35,7 +35,7 @@ jobs:
       - name: Installing packages
         run: |      
           python3.8-dbg -c 'import sys; print("Python debug build:", hasattr(sys, "gettotalrefcount"))'
-          python3.8-dbg -m pip install --upgrade pip 'setuptools<58.5' wheel
+          python3.8-dbg -m pip install --upgrade pip setuptools wheel
```
Comment (mckib2) [inline .github/workflows/linux.yml]: This looks unintentional

File: azure-pipelines.yml [hunk 3]
```diff
@@ -172,7 +172,7 @@ stages:
             curl https://bootstrap.pypa.io/get-pip.py -o get-pip.py && \
             python3.8 get-pip.py && \
             pip3 --version && \
-            pip3 install 'setuptools<58.5' wheel --user && \
+            pip3 install setuptools wheel --user && \
```
Comment (mckib2) [inline azure-pipelines.yml]: This file also seems to include unintentional changes (not sure where they came from...)

File: ci/azure-travis-template.yaml [hunk 4]
```diff
@@ -64,7 +64,7 @@ steps:
   displayName: 'Add ccache to path'
 - script: >-
     pip install --upgrade ${{parameters.numpy_spec}} &&
-    pip install --upgrade pip==20.2.4 'setuptools<58.5' wheel &&
+    pip install --upgrade pip==20.2.4 setuptools wheel &&
```
Comment (mckib2) [inline ci/azure-travis-template.yaml]: Ditto

File: pyproject.toml [hunk 5]
```diff
@@ -10,7 +10,7 @@
 [build-system]
 requires = [
     "wheel",
-    "setuptools<58.5",
+    "setuptools",
```
Comment (mckib2) [inline pyproject.toml]: Ditto

File: scipy/__init__.py [hunk 6]
```diff
@@ -136,7 +136,7 @@ def __dir__():
     from scipy._lib import _pep440
     # In maintenance branch, change to np_maxversion N+3 if numpy is at N
     # See setup.py for more details
-    np_minversion = '1.16.5'
+    np_minversion = '1.17.3'
```
Comment (mckib2) [inline scipy/__init__.py]: Ditto

**Comment by mckib2** [comment]
> Almost. Just review the diff for unintended changes (e.g. there are changes to stats at the bottom that don't look related) first. Then when you post, please add those comments about changes (other than deletions) and why they are necessary. I'd avoid complicating this PR with unnecessary changes.

I figured out what was happening -- apparently I had extra commits in my fork's master branch which was causing "dirty" diffs.  Rebasing everything back onto the upstream master seems to have resolved it
